File: ruby/worker/worker.rb
require_relative '../pipe/http'
require_relative '../lib/retry'

class Worker
  attr_reader :pipe
  def initialize(pipe:, step:) 
    @pipe = pipe
    @step = step
    @running = false

    @on_message_handlers = []
    @on_error_handlers = []
    
    @pipe.receive_options = {
      auto_ack: false,
      block: false,
      count: 10,
      timeout: 60,
      redelivery_timeout: 120
    }
  end

  def receive_options=(auto_ack: nil, block: nil, count: nil, timeout: nil, redelivery_timeout: nil)
    @pipe.receive_options={auto_ack: auto_ack, block: block, count: count, timeout: timeout, redelivery_timeout: redelivery_timeout}
  end

  def on_message(&block)
    @on_message_handlers << block if block
  end

  def on_error(&block)
    @on_error_handlers << block if block
  end

  def stop
    @running = false
  end

  def run
    raise 'already running' if @running
    @running = true
    @pipe.start
    while true
      while pipe.length == 0
        sleep 0.1
      end
      msg = pipe.next
      shouldcomplete = true
      keepon = true
      @on_message_handlers.each do |block|
        break unless keepon
        begin
          block.call(msg)
        rescue => exception
          puts "error on mesage: #{exception.to_s}"
          keepon = false

          Retry.do(40, 0.25) {
            @pipe.log(msg['id']['value'], -1, "failed to complete handler with error #{exception.to_s}")
          }
          if @on_error_handlers.length > 0 
            shouldcomplete = false
          end
          @on_error_handlers.each do |erblock|
            erblock.call(msg, exception)
          end
        end
      end

      if shouldcomplete
        Retry.do(40, 0.250) {
          @pipe.log(msg['id']['value'], 0, "completed step #{@step}")
        }
        Retry.do(40, 0.250) {
          @pipe.complete(msg['id']['value'])
        }
      end
    end
  end
end

def NewHTTPWorker(url:,apikey:,step:)
  pipe = HTTPPipe.new(apikey: apikey, pipe: step, url: url)
  Worker.new(pipe: pipe, step: step)
end

File: js/worker/worker.js
const retry = require('../lib/retry');
const sleep = require('../lib/sleep');
const HTTPPipe = require('../pipe/http');

class Worker {
  constructor(pipe, step) {
    this._pipe = pipe;
    this._step = step;
    this._running = false;
    this._onMessage = [];
    this._onError = [];

    this._pipe.setReceiveOptions({
      autoAck: false,
      block: false,
      count: 10,
      timeout: 60,
      redeliveryTimeout: 120,
    });
  }

  get pipe() {
    return this._pipe
  }

  set receiveOptions(receiveOptions) {
    this._pipe.setReceiveOptions(receiveOptions);
  }

  onMessage(fn) {
    this._onMessage.push(fn);
  }

  onError(fn) {
    this._onError.push(fn);
  }

  stop() {
    this._running = false;
    this.pipe.stop();
  }

  async run() {
    if (this._running) {
      throw new Error('already running');
    }

    this._running = true;

    this._pipe.start()
    await sleep(1000);

    let msg;
    while(msg = await this._pipe.next()) {
      let shouldcomplete = true;
      let keepon = true;
      for (let i = 0; i < this._onMessage.length; i++) {
        const fn = this._onMessage[i];
        if (!keepon) {
          continue;
        }

        try {
          await fn(msg);
        } catch (er) {
          // console.log('threw an error on callback', er)
          keepon = false;
          await retry( async () => {
            await this._pipe.log(msg.id.value, -1, `failed to complete handler ${i}, with error ${er}`);
          }, 40, 250);
          
          if (this._onError.length > 0) {
            shouldcomplete = false
          }
          for (let j = 0; j < this._onError.length; j++){
            await this._onError[j](m, er);
          }
        }
      }
      if (shouldcomplete) {
        await retry( async () => {
          await this._pipe.log(msg.id.value, 0, `completed step ${this._step}`);
        }, 40, 250);
        await retry( async () => {
          await this._pipe.complete(msg.id.value);  
        }, 40, 250);
      }
    }
  }
}

const NewHTTPWorker = (url, apikey, step) => {
  const pipe = new HTTPPipe(apikey, step, url);
  return new Worker(pipe, step);
}

module.exports = {
  Worker,
  NewHTTPWorker,
}